pkg/deployment/ipam.go

```diff
@@ -31,74 +32,64 @@ import (
-			dataplanev1.NodeSetIPReservationReadyErrorMessage)
-		return nil, false, err
-	}
-
-	if len(allIPSets) == 0 {
-		return nil, true, nil
-	}
-
-	for _, s := range allIPSets {
-		if s.Status.Conditions.IsFalse(condition.ReadyCondition) {
-			instance.Status.Conditions.MarkFalse(
-				dataplanev1.NodeSetIPReservationReadyCondition,
-				condition.RequestedReason, condition.SeverityInfo,
-				dataplanev1.NodeSetIPReservationReadyWaitingMessage)
-			return nil, false, nil
-		}
-	}
-	instance.Status.Conditions.MarkTrue(
-		dataplanev1.NodeSetIPReservationReadyCondition,
-		dataplanev1.NodeSetIPReservationReadyMessage)
-	return allIPSets, true, nil
-}
-
-// DataplaneDNSData holds information relevant to the OpenStack DNS configuration of the cluster.
-type DataplaneDNSData struct {
-	// Err holds any errors returned from the Kubernetes API while retrieving data.
-	Err error
-	// AllIPs holds a map of all IP addresses per hostname.
-	AllIPs map[string]map[infranetworkv1.NetNameStr]string
-	// Hostnames is a map of hostnames provided by the NodeSet to the FQDNs
-	Hostnames map[string]map[infranetworkv1.NetNameStr]string
-	// CtlplaneSearchDomain is the search domain provided by IPAM
-	CtlplaneSearchDomain string
+type DNSDetails struct {
+	// IsReady has DNSData been reconciled
+	IsReady bool
 	// ServerAddresses holds a slice of DNS servers in the environment
 	ServerAddresses []string
 	// ClusterAddresses holds a slice of Kubernetes service ClusterIPs for the DNSMasq services
 	ClusterAddresses []string
-	// Ready indicates the ready status of the DNS service
-	Ready bool
+	// CtlplaneSearchDomain is the search domain provided by IPAM
+	CtlplaneSearchDomain string
+	// Hostnames is a map of hostnames provided by the NodeSet to the FQDNs
+	Hostnames map[string]map[infranetworkv1.NetNameStr]string
+	// AllIPs holds a map of all IP addresses per hostname.
+	AllIPs map[string]map[infranetworkv1.NetNameStr]string
+}
+
+// CheckDNSService checks if DNS is configured and ready
+func checkDNSService(ctx context.Context, helper *helper.Helper,
+	instance client.Object, dnsDetails *DNSDetails,
+) error {
+	dnsmasqList := &infranetworkv1.DNSMasqList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(instance.GetNamespace()),
+	}
+	err := helper.GetClient().List(ctx, dnsmasqList, listOpts...)
+	if err != nil {
+		return err
+	}
+	if len(dnsmasqList.Items) == 0 {
+		util.LogForObject(helper, "No DNS control plane service exists yet", instance)
+		return nil
+	}
+	if !dnsmasqList.Items[0].IsReady() {
+		util.LogForObject(helper, "DNS control plane service exists, but not ready yet ", instance)
+		return nil
+	}
+	dnsDetails.ClusterAddresses = dnsmasqList.Items[0].Status.DNSClusterAddresses
+	dnsDetails.ServerAddresses = dnsmasqList.Items[0].Status.DNSAddresses
+	return nil
 }
 
 // createOrPatchDNSData builds the DNSData
-func (dns *DataplaneDNSData) createOrPatchDNSData(ctx context.Context, helper *helper.Helper,
```

Comment:
Can't this stay as a method of DataplaneDNSData? You still have to use the same fields and the structure remains, it's just inflating size of the patch.